src/playground.rs

```diff
@@ -0,0 +1,275 @@
+    let edition = args.params.get("edition").unwrap_or_else(|| &"2018");
+
+    let mut request = PlaygroundCode::new(code);
+
+    match Channel::from_str(channel) {
+        Ok(c) => request.channel = c,
+        Err(e) => errors += &format!("{}\n", e),
+    }
+
+    match Mode::from_str(mode) {
+        Ok(m) => request.mode = m,
+        Err(e) => errors += &format!("{}\n", e),
+    }
+
+    match Edition::from_str(edition) {
+        Ok(e) => request.edition = e,
+        Err(e) => errors += &format!("{}\n", e),
+    }
+
+    if !code.contains("fn main") {
+        request.crate_type = CrateType::Library;
+    }
+
+    let resp = args
+        .http
+        .post("https://play.rust-lang.org/execute")
+        .json(&request)
+        .send()?;
+
+    let result: PlayResult = resp.json()?;
+
+    let result = if result.success {
+        result.stdout
+    } else {
+        result.stderr
+    };
+
+    Ok(if result.len() + errors.len() > 1994 {
+        format!(
+            "{}Output too large. Playground link: {}",
+            errors,
+            get_playground_link(args, code, &request)?
+        )
+    } else if result.len() == 0 {
+        format!("{}compilation succeded.", errors)
+    } else {
+        format!("{}```{}```", errors, result)
+    })
+}
+
+fn get_playground_link(args: &Args, code: &str, request: &PlaygroundCode) -> Result<String> {
+    let mut payload = HashMap::new();
+    payload.insert("code", code);
+
+    let resp = args
+        .http
+        .post("https://play.rust-lang.org/meta/gist/")
+        .header(header::REFERER, "https://discord.gg/rust-lang")
+        .json(&payload)
+        .send()?;
+
+    let resp: HashMap<String, String> = resp.json()?;
+    debug!("gist response: {:?}", resp);
+
+    resp.get("id")
+        .map(|id| request.url_from_gist(id))
+        .ok_or_else(|| "no gist found".into())
+}
+
+pub fn run(args: Args) -> Result<()> {
+    let code = args
+        .params
+        .get("code")
+        .ok_or("Unable to retrieve param: query")?;
+
+    let result = run_code(&args, code)?;
+    api::send_reply(&args, &result)?;
+    Ok(())
+}
+
+pub fn help(args: Args) -> Result<()> {
+    let message = "Compile and run rust code. All code is executed on https://play.rust-lang.org.
+```?play mode={} channel={} edition={} `code` ```
+Optional arguments:
+    \tmode: debug, release (default: debug)
+    \tchannel: stable, beta, nightly (default: nightly)
+    \tedition: 2015, 2018 (default: 2018)
+    ";
+
+    api::send_reply(&args, message)?;
+    Ok(())
+}
+
+pub fn err(args: Args) -> Result<()> {
+    let message = "Missing code block. Please use the following markdown:
+\\`\\`\\`rust
+    code here
+\\`\\`\\`
+    ";
+
+    api::send_reply(&args, message)?;
+    Ok(())
+}
+
+pub fn eval(args: Args) -> Result<()> {
+    let code = args
+        .params
+        .get("code")
+        .ok_or("Unable to retrieve param: query")?;
```

Comment:
I think we can use `ok_or_else` here as well

Reply:
its just a str literal, i don't think it makes any difference for things that aren't function calls